### DISCUSSION TITLE
How to use renderPage() to render a specific page without redirecting in VPSv1?

### DISCUSSION THREAD

**Discussion by Jindorf** [vikejs/vike#925] [Discussion] [answered]
Hi, 

I am using VPSv1 and Express 5 and I am trying to render the Error-Page in `/_error/+Page.vue` on initial Request.

I've tried overwriting `urlOriginal` to an invalid URL aswell as setting `redirectTo` to an invalid URL, but nothing happens, resulting in rendering `/index/+Page.vue`, even though the parameters are submitted to the pageContext.

For a little more context, I have a middleware that catches all Requests to my Express-Server, returning a promise.
If the Request is valid, VPS renders the requested Page.
If the Request is valid, but an invalid sub-page is being called e.g. `.../myValidURL/invalidSubPage`, VPS renders `/_error/+Page.vue`.
If the Request is invalid, my Error-Handling middleware gets called. I want VPS to render  `/_error/+Page.vue` in this case. However, I can't seem to figure it out and it renders the  `/index/+Page.vue

Thanks for helping!

**Comment by brillout** [comment]
Have you seen https://vite-plugin-ssr.com/RenderErrorPage?

**Reply by Jindorf** [reply]
Thanks for the quick reply! 

Yes, I have seen the RenderErrorPage docs and this is how i approached it before migrating to v1. 

But after migrating to v1, onBeforeRender slowed down the app significantly, even if it is not doing anything.
Changing the page (clientRouting is active) takes ~800ms to 1s, on every Request. Without an +onBeforeRender.ts the page loads almost instantly while client-routing. It seems like the server is always involved when onBeforeRender is called, even though it should be server-side only by default iirc. I also tried to explicitly set the env to `server-only` in my `config.h.ts`, but without success.

**Reply by brillout** [reply]
I wouldn't why that happens. I'm inclined to think it's a user land problem you missed? Minimal reproduction welcome if you believe it's a vite-plugin-ssr problem.

**Reply by Jindorf** [reply]
The problem seems to be that I'm doing some async requests before rendering the application in conjunction with a `+onBeforeRender.ts` file being present.

Reproduction: https://github.com/Jindorf/VPSv1-onBeforeRender

As you can see in the example, `+onBeforeRender` is empty.
If you switch between Home and About, the network-tab always shows a request for index.pageContext.json. In this case about 40ms, my main application fetches a config-file that is way bigger and necessary on every page, loading in ~800ms-1000ms per route-switch.
(Moving the data-fetching to `+onBeforeRender` (where it should be imo) didn't improve the performance)
Here is a screenshot of the network tab with an empty `+onBeforeRender`
![image](https://github.com/brillout/vite-plugin-ssr/assets/21169298/b0628a63-ba7c-41cf-b035-0fa2424135cb)

If you now delete the +onBeforeRender file and restart the dev server, everything is loading way faster and the request in the network-tab is no longer there.

What am I doing wrong here? Or is this a bug in vps?

**Reply by brillout** [reply, marked as answer]
This all seems normal to me. I'm not sure what you expect from VPS.

If you define a `onBeforeRender()` hook then it'll always be called. (You can change its `env` to `'server-and-client'` so that, upon subsequent client-side navigation, it's called on the client-side instead of doing an HTTP request.)

**Reply by Jindorf** [reply]
Setting `onBeforeRender` to `server-and-client` worked for me. Thanks alot!
(Before migrating to vpsv1 I defined `onBeforeRender` in `default.page.ts` and didn't see the `onBeforeRender` property in the config, oops.. )

**Reply by brillout** [reply]
👍 Btw. in case you think you can convince your company: https://github.com/sponsors/brillout.
